### PR TITLE
Addition of menu bar; separation of minion vs keys

### PR DESCRIPTION
**Pull request by erwindon**
This PR introduces the menubar as proposed in #20.
The first 2 pages are actually more-or-less the original home page, but split by the minion vs keys functionality.
The original (single) home page is renamed to "minions" page. This one now again contains only entries for active minion-keys. A second page contains the "keys" functionality. This lists all known keys, (i.e.: accepted, rejected, unaccepted and denied keys). Each page shows only the commands (in the dropdown menu) that match the purpose of the page.
The menu-bar code allows pages (such as job details) without corresponding menu-item.
The menu-bar keeps the menu item for the active page highlighted.

### CODE REVIEW

**Comment by maerteijn** [comment]
My first impression is: Looks good! I think the majority of the pull request is about moving code around because you added a extra page right?

PS, could you rebase this branch on master (conflicts)

**Comment by erwindon** [comment]
> I think the majority of the pull request is about moving code around because you added a extra page right?
Yes, and the fact that there are now multiple pages.

The branch has been rebased as needed/requested.

On re-test, I noticed that navigating between the pages still seems to have a timing problem. When you navigate while one/more api-calls is/are still running, then screen becomes empty. It can then only be restored using full page refresh.
I'll look into that first...

**Comment by erwindon** [comment]
The actual problem is cache-control. Under certain circumstances, an API call returns the HTML from the browser cache. That is now prevents.
The PR is valid again.

File: saltgui/static/scripts/routes/keys.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    container.appendChild(element);
+  }
+
+  _addPreMinion(container, hostname) {
+    var element = document.createElement('li');
+
+    element.appendChild(Route._createDiv("hostname", hostname));
+
+    var pre = Route._createDiv("unaccepted", "unaccepted");
+    pre.id = "status";
```
Comment (maerteijn) [inline saltgui/static/scripts/routes/keys.js]: Adding multiple elements with the same id is not allowed in the dom. So you should add a class here. See also my general comments about this pull request

File: saltgui/static/stylesheets/main.css [hunk 2]
```diff
@@ -71,6 +71,22 @@ h1 {
   cursor: pointer;
 }
 
+.menu_item {
+  display: inline-block;
+  padding-left: 20px;
+  font-size: 20px;
```
Comment (maerteijn) [inline saltgui/static/stylesheets/main.css]: Can you change this to `18px` ?

**Comment by maerteijn** [comment]
I think this is a change for the better (and the future) but there are some elements I'm not so keen on (yet). Let's take the following example:

key-duplicates

- When you reject a key, the page will quickly reload reload but still shows the same key as accepted here. This is not the case when you completely reload the page. 
- I don't see why it displays 'rejected' next to ``8be47cba35e2`` when it's already listed under 'rejected keys'. See the comment here below.
- The layout with different fields (os description in the accepted keys list, no description in the others) becomes a bit of a mess. Wouldn't it be a better idea to do something like:
menu-suggestion

Note that you use the same class for the `status` and the `os` display to make sure the button alignment stays a bit sane.

File: saltgui/static/scripts/dropdown.js [hunk 3]
```diff
@@ -17,6 +18,7 @@ class DropDownMenu {
     var button = Route._createDiv("run-command-button", title);
     button.addEventListener('click', evt => callback(evt));
     this.menuDropdownContent.appendChild(button);
+    this.menuDropdown.style.display = "block";
```
[the remark this answers is not in the document]
Comment (erwindon) [reply saltgui/static/scripts/dropdown.js]: done

File: saltgui/static/stylesheets/main.css (hunk 2 above)
Comment (erwindon) [reply saltgui/static/stylesheets/main.css]: done (I assume you meant the font-size)

**Comment by erwindon** [comment]
Strange, the page reload after a successful key-management api call does a `location.reload()`, which is the same a complete page reload. The problem seems to be (again) a timing error. We use 2 api calls: 1 to get the keys, and 1 to get the grains info. In your case, the effect can only be reproduced when the grains info returns faster than the keys info. Because the grains info is still returned, an item is added to the "accepted" block. Because the keys info says the key is rejected, an item is added to the "rejected" block. therefore ending with 2 items on the screen. I'll make the keys information leading on page "keys" and remove any accidental successful grains information.
please wait...

**Comment by erwindon** [comment]
I have a lot of offline minions (average 50%) so that effect is less visible here. The denied/unaccepted/rejected sections are mostly empty. The other pages that I'm proposing also have the same layout with both status/ip and os. Can we postpone this layout change to a bit later?

**Comment by maerteijn** [comment]
Ok, I'm willing to postpone the layout changes later. The only thing that really is a concern for me is this:

div layout

The accepted keys section has three columns, the other one two. Could you at least have the same amount of columns (with the same css width etc) to have the buttons aligned equally?

**Comment by erwindon** [comment]
Finally, I understand... In the resolution that I used, the command buttons were floating-right. But not in the lower resolutions. Then indeed all columns must be present (easy fix).

**Comment by erwindon** [comment]
I think I've updated all the requested changes.
Note that I treated it as the original PR, i.e. it is a PR only for branch `pages3`.
Our common effort on feature `hamburger` is not included here.
That branch now needs a rebase and should be in its own merge (with or without a PR).

**Comment by maerteijn** [comment]
Fantastic work with the fixes, just tested and looks fine!

**Pull request merged**